Purpose: Make sure the .depend.* files have complete dependencies when
         recursive virtual facilities are used.
Fixes:   #534526
Status:  From upstream.
---
Index: insserv/insserv.c
===================================================================
--- insserv.orig/insserv.c	2009-09-26 22:35:39.000000000 +0200
+++ insserv/insserv.c	2009-09-26 22:35:42.000000000 +0200
@@ -2109,33 +2109,26 @@
 	    goto out;
 	}
 
-	if ((*deep)++ > 10) {
-	    warn("The nested level of the system facilities in the insserv.conf file(s) is to large\n");
-	    goto out;
-	}
-
 	list_for_each_safe(tmp, safe, ptr) {
 	    repl_t * rnxt = getrepl(tmp);
 	    if (*rnxt->r[0].name == '$') {
-		expand_faci(tmp, head, deep);
-	    } else {
-		if (*deep == 1) {
-		    if (--(*rent->r[0].ref) <= 0)
-			free(rent->r[0].ref);
-		    rent->r[0] = rnxt->r[0];
-		    ++(*rent->r[0].ref);
-		} else {
-		    repl_t *restrict subst;
-		    if (posix_memalign((void*)&subst, sizeof(void*), alignof(repl_t)) != 0)
-			error("%s", strerror(errno));
-		    insert(&subst->r_list, head);
-		    subst->r[0] = rnxt->r[0];
-		    ++(*subst->r[0].ref);
+		if (*deep > 10) {
+		    warn("The nested level of the system facilities in the insserv.conf file(s) is to large\n");
+		    goto out;
 		}
+		(*deep)++;
+		expand_faci(tmp, head, deep);
+		(*deep)--;
+	    } else if (*deep > 0) {
+		repl_t *restrict subst;
+		if (posix_memalign((void*)&subst, sizeof(void*), alignof(repl_t)) != 0)
+		    error("%s", strerror(errno));
+		insert(&subst->r_list, head->prev);
+		subst->r[0] = rnxt->r[0];
+		(*subst->r[0].ref) = 1;
 	    }
 	}
 out:
-	(*deep)--;
 	return;
 }
 
